### DISCUSSION TITLE
Next-gen Extractor Project

### DISCUSSION THREAD

**Discussion by RimWorldKorea** [RimWorldKorea/RMK#394] [일반]
#223 -> 이어서..

**Comment by csh1668** [comment]
[net7.0-windows.zip](https://github.com/RimWorldKorea/RMK/files/13888322/net7.0-windows.zip)
20240110/22:40 버전
    /*
     * 변경사항:
     *
     * Internal
     * - NodeReplacement 구현: 예시) XML 변환 과정에서 ScenarioDef+label를 ScenarioDef+scenario.name로 바꿔서 변환해줌
     * - Full-list translation 구현
     * - Keyed 추출 관련: (파일이름).(노드이름)의 형식에서 (파일이름)|(노드이름)의 형식으로 변경.
     *   이때 파일이름은 선택사항이며, 알파추출기 형식처럼 없어도 됨.
     * - Pointer 구현: 참조를 원하는 Identifier의 이름이 `RandomDef+hello.label` 이라면 번역문에 `{*RandomDef+hello.label}`처럼 적습니다.
     * - 엑셀 -> XML 변환 관련: 알파 추출기에서 추출한 엑셀 파일을 먹여도 잘 작동하게 변경 (테스트는 안해봄)
     * - LoadFolders로 추가되는 노드도 RequiredMod에 표시 관련: LoadFolders에 있는 모드 정보는 packageId인데
     *   표시해야 하는 정보는 모드 이름이므로, packageId가 일치하는 모드를 사용자가 구독한 모드 중에서 찾음.
     *   못찾을 경우 '##packageId##'라는 특수한 접미사가 붙게 되며, 이는 사용자가 직접 수정해야 함.
     *
     * GUI:
     * - 참조 모드 기본값 설정을 저장, 불러오기 가능: 우클릭 후 저장 or 불러오기
     * - 설정 구현
     * - 선택한 참조 모드, 추출 모드만 보는 필터 기능
     * - XLSX -> XML 기능 구현
     *
     */

Keyed 관련 이유로 콘솔 때랑은 호환이 안될 수 있어요

**Reply by csh1668** [reply]
구상중인 기능
- 설정에 참조 모드 기본 설정을 빼놔서 추출 모드 선택할 때 자동으로 선택되도록하기
- 선택한 모드의 modDependencies나 LoadFolder.xml를 잘 파악해서 참조 모드가 자동 선택되도록 하기
- 추출 폴더 선택할때 짤려서 보이는 부분은 마우스를 올리고 있으면 전체 정보가 뜨도록 해서 해결
- LoadFolders에서 IfModActive의 값이 ','가 들어가면서 여러개를 지정할 수 있는데 그거 대응 (예시:ↁ House Sanguin)

**Reply by RimWorldKorea** [reply]
XMLister도 합치면 #395 같은 상황에 대응할 수 있는 기능도 생각해볼 수 있을 것 같아요
중국산 모드 괜찮은거 나오면 번역하는 사람이 종종 있더라구요

요약하면
추출시 원문 대신 특정 언어의 번역 데이터를 원문에 덮어씌워주는 기능

**Reply by RimWorldKorea** [reply]
<img width="640" src="https://github.com/RimWorldKorea/RMK/assets/150018627/9c316cb3-4d39-4b5b-a993-fc028f811b8d"><br>

새 버전 키 노드가 안뽑혀요 😮

LoadFolders로 불러오는 폴더에 있는건 가져오는데

정작 루트 폴더 Languages에 있는게 안나오네요

<img width="463" src="https://github.com/RimWorldKorea/RMK/assets/150018627/10debe8b-004b-4c0a-a068-38d499e4ee93">

**Reply by csh1668** [reply]
확인 완료
그나저나 GUI 생김새가 제가 실행했을때랑 완전 딴판이네요..?
![스크린샷 2024-01-10 223916](https://github.com/RimWorldKorea/RMK/assets/18442452/5b6586be-08db-437d-a0e7-90f4cc31c4e1)

**Reply by csh1668** [reply]
[net7.0-windows.zip](https://github.com/RimWorldKorea/RMK/files/13888322/net7.0-windows.zip)
번거롭게 해서 정말 송구스럽네요 ㅠ
- ' 문자 오타 수정
- 기본 폴더 추출 안하는 문제 수정
- 기본 추출 태그 변경
핫픽스입니다
혹시나 해서 글 가장 위에 것도 업데이트 해놨습니다

https://github.com/RimWorldKorea/RMK/discussions/394#discussioncomment-8078725
중국어 원문 모드는 설정에서 원문 언어를 중국어로 바꾸면 해결할 수 있지 않을까요?

**Reply by RimWorldKorea** [reply]
xlsx의 원문 추출 텍스트를 모드의 원본 텍스트가 아니라  
해당 모드에 내장된 다른 언어 데이터로 덧씌우고 싶은데  
이걸 추출 단계에서 스마트하게 해줬으면 한다는 뜻이예요

기능이 없어도 가능은 한 일이지만 아마 대부분 유저들은 못할거라..

중요도가 높은건 아니니 지금은 이런 소요가 있다는 것만 기억해두면 될 것 같아요
<br>

화면이 다른건 아마 윈도우 차이가 아닐까요? 아직 10을 쓰고 있어서 ㅋㅋ

**Reply by csh1668** [reply]
추출 폴더 선택할때 두번째줄에 요구 모드도 같이 떠야되는데 안뜨네요.. 뭐지

**Reply by RimWorldKorea** [reply]
이제 보니깐 의존성 부분이 잘렸네요?

자세히 보니깐 표시는 되는데 칸 높이가 낮아서 안보이는 것 같아요

**Reply by RimWorldKorea** [reply]
이번 버전에선 오피셜이랑 로컬 모드가 안잡혀요~

<img width="560" src="https://github.com/RimWorldKorea/RMK/assets/150018627/a18aaa7e-f586-43ce-a7b6-057dd62e8799">

**Reply by csh1668** [reply]
![스크린샷 2024-01-13 125006](https://github.com/RimWorldKorea/RMK/assets/18442452/7cb0ef6c-5503-447f-8997-d666f7ca037d)
![스크린샷 2024-01-13 125028](https://github.com/RimWorldKorea/RMK/assets/18442452/ab51ce83-578c-4ea9-a4e4-3f08fbbcc6b6)
경로 지정 잘 되어있는거 맞죠?
그 부분은 건든게 없어서 아마 문제 없을텐데..

라고 생각해보니 설정이랑 경로 관련해서 하나 문제가 있는데
[net7.0-windows.zip](https://github.com/RimWorldKorea/RMK/files/13926605/net7.0-windows.zip)
그거 수정한 버전입니다


이놈의 프로그래밍은 버그가 끝없네요 아이고 ㅠㅠ

**Comment by RimWorldKorea** [comment]
<img width=480 src="https://github.com/RimWorldKorea/RMK/assets/150018627/619bc103-175f-497a-a26c-a89521c698fa"><br>

에러 케이스에 따라 다른 메시지를 띄우면 좋을 것 같아요
완료가 아니야~

**Reply by csh1668** [reply]
에러가 떴긴 했지만 완료된 건 맞거든요.. ㅋㅋㅎㅎ

**Comment by RimWorldKorea** [comment]
기본 배포 추출 태그는 이걸로!

    label/rulesStrings/description/baseDesc/title/titleShort/customLabel/symbol/jobString/reportString/labelNoun/slateRef/verb/gerund/adjective/member/tips/ideoName/thoughtStageDescriptions/jobReportString/theme/labelShortAdj/labelPlural/letterText/deathMessage/labelShort/letterLabel/helpText/text/baseInspectLine/labelFemale/descriptionShort/beginLetter/ingestCommandString/ingestReportString/titleShortFemale/titleFemale/gerundLabel/pawnLabel/stageName/shortDescription/customEffectDescriptions/endMessage/leaderTitle/pawnSingular/pawnsPlural/desc/recoveryMessage/chargeNoun/cooldownGerund/type/potentialExtraOutcomeDesc/labelNounPretty/headerTip/rejectInputMessage/spectatorGerund/spectatorsLabel/fuelLabel/formatString/useLabel/RMBLabel/permanentLabel/name/missingDesc/worshipRoomLabel/labelAbstract/fuelGizmoLabel/destroyedLabel/outOfFuelMessage/summary/ritualExpectedDesc/customSummary/meatLabel/labelForFullStatList/tooltip/gizmoLabel/onMapInstruction/letterTitle/textEnemy/destroyedOutLabel/beginLetterLabel/labelMale/groupName/gizmoDescription/names/arrivalTextEnemy/letterLabelEnemy/arrivedLetter/calledOffMessage/finishedMessage/approachingReportString/approachOrderString/expectedThingLabelTip/skillLabel/extraPredictedOutcomeDescriptions/modNameReadable/descriptionFuture/textWillArrive/arrivalTextFriendly/letterLabelFriendly/helpTextController/successfullyRemovedHediffMessage/textFriendly/eventLabel/textController/descOverride/shortDescOverride/content/discoveredLetterText/discoveredLetterTitle/beginLetterContinue/resourceLabel/message/overrideLabel/extraTooltip/offMessage/successMessage/effectDesc/letterInfoText/categoryLabel/groupLabel/battleStateLabel/customizationTitle/fixedName/noun/lockedReason/descriptionExtra/labelPrefix/labelMechanoids/ingestReportStringEat/failMessage/valueFormat/structureLabel/labelSocial/labelInBracketsExtraForHediff/ChooseDesc/ChooseLabel/ritualExplanation/resourceDescription/discoverLetterText/countdownLabel/inspectString/completedLetterText/completedLetterTitle/leaderDescription/formatStringUnfinalized/jobReportOverride/discoverLetterLabel/instantlyPermanentLabel/notifyMessage/onCooldownString/invalidTargetPawn/noAssignablePawnsDesc/reportText/statLabel/visualLabel/commandDescriptions/successMessageNoNegativeThought/tipLabelOverride/mainPartAllThreatsLabel/customChildDisallowMessage/ritualExpectedDescNoAdjective/loweredName/cancelLabel

**Comment by RimWorldKorea** [comment]
<img width=240 src="https://github.com/RimWorldKorea/RMK/assets/150018627/27d528c4-1ee9-4de1-a583-d47f57b202e8"><br>

끝에 `'`는 무언가 표시하는 건가요?

**Reply by csh1668** [reply]
그건 오타 ㅋㅋ 다음 버전에서는 지워질 예정인데
다음버전에서 사용하실려면 그거 지우셔야 할 거에요

**Comment by RimWorldKorea** [comment]
추출시 Keyed 주소값에는 출처 파일 정보를 넣지 않는게 기본값인게 좋을 것 같아요

어느 파일에 들어있는지는 Keyed보단 오히려 DefInjection 쪽이 필요한데

그냥 윈도우 파일 검색기 돌리면 되기도 하는데다

알파 추출기로 작업했던 파일을 전환할 때 수고가 늘 것 같더라구요

**Reply by RimWorldKorea** [reply]
**1.**

<img width=420 src="https://github.com/RimWorldKorea/RMK/assets/150018627/95ee1d25-e784-4005-84e0-a0eb3160d423" />
<br><br>

XSLX → XML 기능 사용 시 열려있는 파일을 참조할 경우 오류창이 뜨면서 강제 종료하게 만드는데  
그냥 작업을 취소시킨다던지 해서 강제 종료하지 않도록 하면 깔끔할 것 같아요

알파추출기 같은 경우는 어떻게 했는지 열린 파일에도 돌아가더라구요

**2.**

<img width="260" src="https://github.com/RimWorldKorea/RMK/assets/150018627/b1c70bf7-b262-4a4e-9c8a-78db06ad740e">
<br><br>

XSLX → XML 기능 사용 시 번역문을 넣지 않은 노드는 생략하는 기능이 필요해요! <- _중요_

**Reply by RimWorldKorea** [reply]
**3.**

<img width="704" alt="화면 캡처 2024-01-17 133628" src="https://github.com/RimWorldKorea/RMK/assets/150018627/cb71ed3d-f4bf-414d-93dc-0ff1bd9c5036">
<br><br>

노드 대체 기능이 작동하지 않아요

**Reply by RimWorldKorea** [reply]
**4.**

<img width=560 src="https://github.com/RimWorldKorea/RMK/assets/150018627/34fe2b1f-8c9a-478c-8568-cba64f5cb0fb" />
<br><br>

프로그램을 열고 바로 `추출할 모드 선택`을 누르면 모드 목록이 뜨지 않고
반드시 `옵션`창을 한 번 열어야 정상적으로 목록이 불러와져요

**Comment by csh1668** [comment]
추출시 Keyed 주소값에는 출처 파일 정보를 넣지 않는게 기본값인게 좋을 것 같아요
알파 추출기로 작업했던 파일을 전환할 때 수고가 늘 것 같더라구요

- XML로 변환 시, Keyed의 Node값에는 파일 정보가 있어도 되고 없어도 되기 때문에 굳이 일일이 파일정보를 넣으며 수정 안하셔도 됩니다!

1, 2, 4
- 피드백 확인했습니다

3
- 노드 대체 기능은 XML로 변환하는 단계에서만 적용됩니다. 엑셀 단계에서는 바꾸지 않고 XML 변환과정에서 바꾸는 것이, 알파추출기에서 추출한 엑셀 파일도 작업자가 일일이 수정하지 않고도 노드대체 기능을 적용할 수 있어서 더 효과적이라 생각합니다

**Reply by RimWorldKorea** [reply]
Keyed에 파일 정보를 빼달라고 하는 이유가..

**1. Keyed 파일의 출처 파일 정보를 활용할 수 있는 영역이 없어요!**
주소 자체는 Keyed 폴더가 아니라 어셈블리에 있는거고  
번역 적용도 _파일 구분 자체에 종속적인 부분이 없기 때문에_  
출처 파일 정보가 유의미한 역할을 하지 않아요

**2. 그런데 그게 주소에 끼어있으면 기존 번역을 새 서식으로 옮기는데 번거로움이 많아요!**
예를 들어서 기존 번역 자료를 XMLister로 가져온 뒤 새 서식에 매핑하기 위해선  
기존 데이터의 주소를 새로 추출한 주소에 대응시키는 과정에서
새로 추출한 주소의 파일 정보를 제거한 테이블을 만들고 거기에 대해서 이중으로 매칭하는 단계가 필요한데
파일 정보를 활용할 수 있는 부분이 없는 점과 더불어 평범한 사용자들의 엑셀 활용 능력을 고려해보면
그냥 빼버리는 편이 작업 흐름을 더 매끄럽게 만들 수 있어요

요약하면..  
Keyed 파일 정보가 필요한 곳이 있는가? X
Keyed 파일 정보가 있으면 일이 늘어나는가? O
정도로 정리할 수 있을 것 같아요

**Comment by csh1668** [comment]
[net7.0-windows.zip](https://github.com/RimWorldKorea/RMK/files/14000508/net7.0-windows.zip)

피드백 다 반영했고 기능 몇 개 추가했습니다

- 설정에 기본 참조모드 리스트 경로 지정 항목 추가
- 추출할 폴더를 고르는 ListBox에 마우스를 가져다대면 전체 텍스트를 보여주도록 변경
- 더 세밀한 Patches 지원: PatchOperationAdd에서 xpath가 Defs고 value로 완전한 Def을 추가하는 경우 지원 (Combat Extended에서 빠뜨린 텍스트 추출 가능 10858 -> 11255)

**Reply by RimWorldKorea** [reply]
<img src="https://github.com/RimWorldKorea/RMK/assets/150018627/b4f5aa98-0269-442e-859e-a583f85b4338" width=480 />
<br><br>

사진으로는 안보이는데 텍스트 박스가 엄청난 속도로 깜빡이면서 동작해요 ㅋㅋ  
뭔가 갱신 주기 같은게 너무 짧은건가?

의존성 부분이 없는 라인은 정적인 상태로 잘 표시돼요

**Reply by RimWorldKorea** [reply]
<img width=360 src="https://github.com/RimWorldKorea/RMK/assets/150018627/ca225673-8f1a-4b4e-b187-905f9f93c084">
<img width=360 src="https://github.com/RimWorldKorea/RMK/assets/150018627/e379dbd0-27f9-40cf-b21f-d6610de9a678">
<br><br>

xlsx->xml 변환 과정에서
String의 번역문이 채워져있지 않으면 원문으로 채워주는 기능이 필요해요
이거 번역한 것만 뽑아서 넣어두면 저것만 나왔던 걸로 기억해요

**Reply by RimWorldKorea** [reply]
<img width= 360 src="https://github.com/RimWorldKorea/RMK/assets/150018627/36ce8854-1357-47c6-9f1d-85f5e7cf7d37">
<br><br>

xlsx->xml 변환시 xml 선언문이 빠져있어요!
`<?xml version="1.0" encoding="utf-8"?>`

**Reply by RimWorldKorea** [reply]
노드 대체 기능이 xlsx->xml 과정에 적용되는 것 뿐 아니라  
추출 시점에서 적용되는 것도 필요할 것 같아요

바익쪽에 복잡한게 하나 있는데
이게 번역을 막 해버리면 동작이 맛이 가는거라..

예를 들어

    <ThingDef ParentName="BaseHumanMakeableGun">
    <defName>VFED_Gun_ChargeCycler</defName>
    <verbs>
      <li>
        <verbClass>Verb_Shoot</verbClass>
        <label>Single shot (Charged shot)</label>
      </li>
      <li>
        <label>Single shot (Thump shot)</label>
      </li>
    </verbs>
    <comps>
      <li Class="MVCF.Comps.CompProperties_VerbProps">
        <compClass>MVCF.Comps.Comp_VerbProps</compClass>
        <verbProps>
          <li>
            <label>Single shot (Charged shot)</label>
          </li>
          <li>
            <label>Single shot (Thump shot)</label>
          </li>
        </verbProps>
      </li>
    </comps>
  </ThingDef>

이런 Def가 있는데

compClass가 `MVCF.Comps.Comp_VerbProps`인 comp가 있으면
해당 comp의 `\verbProps\li\label`과 값이 동일한 해당 Def의 `\verbs\li\label`을 찾아서
번역자가 번역할 수 없도록 xlsx에 포함되지 않도록 한 뒤
`\comps\li\verbProps\li\label`을 `\comps\li\verbProps\li\visualLabel`로 바꿔줘야 해요

이걸 xlsx->xml 과정에서도 충분히 처리할 수 있겠지만
xlsx 파일에 실제로 적용되는 형태의 노드와 그렇지 않은게 섞여있으면 번역자에게 혼란이 생길 수 있겠더라구요

**Reply by RimWorldKorea** [reply]
Translation Handle을 적용할 때

현재 적용된 세팅으로는 `MVCF.Comps.Comp_VerbProps` → `MVCFCompsComp_VerbProps` 이런 식으로 되는데
실제론 `MVCF.Comps.Comp_VerbProps` → `Comp_VerbProps` 이렇게 되어야 해요

마지막 구두점 앞은 다 날리는 것 같긴한데..  
정확한 규칙이 어디 박혀있는지 도통 찾을 수가 없네요 😪

**Reply by RimWorldKorea** [reply]
<img src="https://github.com/RimWorldKorea/RMK/assets/150018627/68dedccd-7d27-4b80-a0d5-ec672297a695" width=960>
<br><br>

요런 식으로 각 설정이 어느 부분에서 적용되는지 명확히 알 수 있도록
UI 상에서 구분하면 좋을 것 같아요

**Reply by csh1668** [reply]
[net7.0-windows.zip](https://github.com/RimWorldKorea/RMK/files/14001532/net7.0-windows.zip)

https://github.com/RimWorldKorea/RMK/discussions/394#discussioncomment-8196339
https://github.com/RimWorldKorea/RMK/discussions/394#discussioncomment-8196653
https://github.com/RimWorldKorea/RMK/discussions/394#discussioncomment-8196729
요 피드백들 반영된 수정버전이에요

**Reply by csh1668** [reply]
https://github.com/RimWorldKorea/RMK/discussions/394#discussioncomment-8197060
TranslationHandle은  ` `, `.`, `\n` 등과 같은 유효하지 않은 문자열이 있으면 안되기 때문에 중간에 한번 Normalize 해주는 작업을 거칩니다.
림월드 뜯어보면서 나온 코드 그대로 넣었으니까 림월드랑 동일한 형식의 TranslationHandle을 생성해줍니다.

근데 위 경우에서는, 그 TranslationHandle에 해당되는 필드의 타입이 Type 타입이라서 실제로는 내부 텍스트인 `MVCF.Comps.Comp_VerbProps`가 `Comp_VerbProps`가 되어야 하네요

근데 이걸 무조건 `.`를 기준으로 쪼개서 마지막껄 택하기도 뭐한게, Type 타입이 아닌 다른 string 타입이거나 다른 타입의 경우에는 내부 텍스트가 그대로 유지되어 `MVCFCompsComp_VerbProps`가 되어야 하기 때문에..... 구현이 매우 어렵다고 볼 수 있겠네요


즉석으로 떠올린 방안으로는 `*compClass`처럼 앞에 특수한 기호가 붙으면 얘는 Type 타입으로 간주하여  `.`을 기준으로 쪼개서 마지막껄 택하도록 하도록 하는 방법일까요..



https://github.com/RimWorldKorea/RMK/discussions/394#discussioncomment-8196915
`\comps\li\verbProps\li\label`을 `\comps\li\verbProps\li\visualLabel`로 바꾸는건 그렇다쳐도

compClass가 `MVCF.Comps.Comp_VerbProps`인 comp가 있으면
해당 comp의` \verbProps\li\label`과 값이 동일한 해당 Def의 `\verbs\li\label`을 찾아서
번역자가 번역할 수 없도록 xlsx에 포함되지 않도록 한 뒤

이건 일반화하기도 무척 어려울뿐더러 구현을 어떻게 해야할지 상상이 안되네요....
아무튼 생각해보는걸로

참고로 노드대체기능은 아직 li가 포함된 녀석은 지원하지 않습니다 <- 요건 수정 예정


https://github.com/RimWorldKorea/RMK/discussions/394#discussioncomment-8197221
엑셀 -> XML 변환 과정  / 모드 -> XML 추출 과정
이 두 과정은 첫 과정을 제외하고 같은 방식으로 처리하고 있습니다.
![제목 없음](https://github.com/RimWorldKorea/RMK/assets/18442452/125ab780-d702-4b6b-8df9-1013e7c5b75a)

그림으로 표현하면 이런 느낌이에요

현재 노드 대체 기능은 번역데이터 -> XML 에서만 작동하고 있는데 이걸 모드 -> 번역데이터 단계에서도 수행하도록 하면 괜찮지 않을까 싶네요

**Reply by RimWorldKorea** [reply]
compClass, verbClass 등 핸들로 사용되는 각각의 요소 안에서는 동일한 규칙을 적용할 수 있다면
각 Translation Handle 태그마다 별도로 규칙을 지정하는게 최선 아닐까요?

**Comment by RimWorldKorea** [comment]
<img width="343" alt="화면 캡처 2024-01-23 214553" src="https://github.com/RimWorldKorea/RMK/assets/150018627/4c358ac5-acaa-424c-85f8-8a4e3891fa33">

xlsx->xml 변환시 파일명에 확장자까지 포함되는데
이건 떼도 될 것 같아요!

**Reply by RimWorldKorea** [reply]
<img width="360" src="https://github.com/RimWorldKorea/RMK/assets/150018627/83309ea3-2d38-4984-bf7e-ef7abe14ed52">
<br><br>

추출 모드 설정창에서 여기도 잘려요

**Comment by RimWorldKorea** [comment]
모드 하나 테스트해보다 알아챈건데
핸들로 쓸 label이 같은 Def 안에서 중복될 경우 이걸 어떻게 할건지 문제가 있네요

예를 들어 리스트 각 요소들의 label이 전부 같아서 핸들로 변환시 구분할 수 없는 경우라면...

일단은 그런 중복이 있는 리스트만 핸들을 쓰지 않도록 할 수도 있겠지만
이러면 뭐는 번호로 돼있고 뭐는 핸들로 돼있고 하는 난잡함이 늘어날 것 같아서

핸들로 다뤄야 편한 소수의 특수 사례를 제외하곤
번호 인덱스를 기본적으로 쓰게 하는게 나은가 고민이 생기네요

**Comment by RimWorldKorea** [comment]
지금 시점에서 추출기에 필요한 고오-급 기능을 정리해보면
최종적으로 모드->xlsx 단계에서 아래 세 가지 상황을 다룰 수 있어야해요

1. OVERRIDE-POP
someScenarioDef\scenario\name이 _정의되어 있지 않으면_
someScenarioDef\label 값으로 someScenarioDef.scenario.name을 정의 후
someScenarioDef.label은 제거

2. FIND-OVERRIDE
someThingDef\verbs\li\verbClass 값이 Verb_Shoot, Verb_ShootOneUse, Verb_ShootWithSmoke 중 하나이고
someThingDef\verbs\li\label이 _정의되어 있지 않으면_
someThingDef\label 값으로 someThingDef.verbs.li.label을 정의

3. FIND-POP-OVERRIDE
someThingDef\comps\li\compClass 값이 MVCF.Comps.Comp_VerbProps이면
someThingDef.comps.li.verbProps.li.label 값과 같은 값을 가지는 someThingDef.verbs.li.label을 제거 후
someThingDef\comps\li\verbProps\li\visualLabel이 _정의되어 있지 않으면_
someThingDef\comps\li\verbProps\li\label 값으로 someThingDef.comps.li.verbProps.li.visualLabel을 정의

이미지로 표현하면 아래와 같이 정리할 수 있을 것 같아요

<img src="https://github.com/RimWorldKorea/RMK/assets/150018627/b2567a7e-6b99-473b-9139-f4e2e04d9dff" width=560 >

**Reply by csh1668** [reply]
그 작업을 위해서는 그냥 노드 대체 기능을 삭제하고 새로운 스크립트 기능을 넣는게 낫겠네요

1번을 예로들면
If (FIND_NODE("ScenarioDef+label") != null&& FIND_NODE("ScenarioDef+scenario.name") == null){ 
CHANGE_CURRENT_NODE("scenario.name");
}
이와같은 스크립트를 작성해볼 수 있겠네요

이 스크립트들은 Defs 추출 과정 중 번역 데이터를 찾았을 때 그 번역 데이터를 return 하기 전에 작동하도록 하여 번역 데이터를 반환 직전에 수정할 수 있습니다. 응용하면 2, 3 기능도 여기에 포함시킬 수 있을 것 같습니다 다만 잘은 모르겠음
아무튼 이 스크립트들은 설정에서 사용자가 자유롭게 추가하고 수정할 수 있게 합니다.
이 방법의 문제점은 사용자가 어느정도 C#에 대한 지식이 있어야 건들 수 있다는 점입니다.
아마 대부분의 번역자분들은 건들지 않고 기본값으로 두고 쓸 기능일거같은데.. ㅋㅋ
차선책으로는 Patches XML과 동일한 문법으로 구현하든지, 파이썬같은 더 쉬운 언어를 사용하는 방법이 있겠네요.


추가로 오늘부터 해외여행이라 다음주까진 개발이 어렵습니다 
피드백 주시는건 계속 읽을 수 있으니까 계속 남겨주세요~

**Reply by RimWorldKorea** [reply]
커스텀 스크립트처럼 너무 범용성있게 만드는 것 보단  
어느정도 틀에 박힌 수준으로 제공하는게 적당할 것 같아요

어차피 사용이 어려우면 아무도 안 쓸거기도 하고
대응해야할 케이스가 앞으로 더 생길까 싶기도 하고
기본 사양에서도 뭔가 더 확장할 일이 있을까 생각해보면
프레임워크 모더들이 뭘 더 복잡한걸 넣지 않는 이상 그럴 일은 없을 것 같더라구요

당장 요구되는 기능만 적당한 인터페이스에 담을 수 있으면 되지 않을까 해요

**Reply by csh1668** [reply]
뭐든지 일반화하고 싶은게 프로그래머 종특이라
각각의 경우에만 대응되는 기능을 만들면 좀 괴롭습니다 ㅎㅎ
근데 이게 쉽긴 하지요 역시

**Comment by RimWorldKorea** [comment]
**1.**
컨텐츠가 `Common` 폴더에 있으면 인식이 안되는 것 같아요
이게 LoadFolders 없이도 기본으로 지원하는 경로였든가 그럴거예요

**2.**
Add 패치가 추출이 안되는 모드를 몇 개 찾았는데
다들 [이 모드](https://steamcommunity.com/sharedfiles/filedetails/?id=2826922787)랑 유사한 구조더라구요
추출기 마지막 버전 기준으로 [여기](https://github.com/RimWorldKorea/RMK/files/14150144/2826922787.xlsx)까지 뽑히는데
패치 부분까지 전부 뽑을 수 있는 규칙이 필요해요

**Comment by RimWorldKorea** [comment]
![제목 없음](https://github.com/RimWorldKorea/RMK/assets/150018627/36d0cda7-9da4-48cb-84c8-99247781efff)

생각해봤는데 https://github.com/RimWorldKorea/RMK/discussions/394#discussioncomment-8239973 예시로 든 것처럼 할 수 있으면 그러는게 역시 나은게 아닌가... 싶은 생각이 들더라구요
적용되는 위치와 우선순위 등도 지정할 수 있도록 유연성 있게 만들어두면
역시 나중이 편한?게 아닌가?

대충 그림과 같은 흐름으로 생각해봤는데 어떻게 보세요?

**Reply by csh1668** [reply]
제가 생각한 내용이 딱 그림 느낌이에요!
우선순위도 지정할 수 있게 해주면 유연하게 할 수 있겠네요

문제는 우선순위는 떨어지는 기능인데 개발 난이도가 너무 높아서
지금은 사실 손 놓고 있습니다..

**Reply by RimWorldKorea** [reply]
그러면 좀 덜 우아하더라도 개발 코스트가 적은 방향이 있다면 그 쪽으로 진행하고
위 방법은 장기적인 챌린지로 설정해두는게 어떨까요?

일단 당장은
- https://github.com/RimWorldKorea/RMK/discussions/394#discussioncomment-8236242 의 1, 2번 항목의 지원
- patch와 loadfolders와 관계된 데이터를 읽어오는 부분을 좀 더 보완

이 된다면 배포를 슬슬 고려해볼 수 있을 시점이 될 것 같아요

**Comment by csh1668** [comment]
/*
     * 작업 내용:
     * Internal)
     * 1. Translation Handle 태그의 앞에 '*'가 붙으면 이는 Type 타입으로 간주하여, 노말라이징을 다르게 수행하도록 변경 (바익 탈영병 MVCF.Comp_VerbProps 관련 문제)
     * 2. PatchOperationAdd 수행 시, xpath로 있는 Def이 참조 모드의 Def이면 추출을 못하는 문제 해결 (바익 페르소나 무기 관련 문제)
     * 3. Common 폴더 고려
     * 4. XMLExtension의 설정이 Patches에 있을 경우도 고려 ([FSF] FrozenSnowFox Tweaks 관련 문제)
     * 5. Patches 자동 생성 방식 개선 (매번 FindMod -> Replace가 아니라, FindMod가 같은 애들끼리는 하나의 Sequence에서 동작하도록)
     * 6. 저장방식 기본값을 Languages XML로 변경 (알파 추출기 기본값과 동일하게 하기 위함)
     * 7. XMLExtention의 tkey나 tTooptip에 접두사로 XMLExtension.이 붙던 문제 수정
     *
     * GUI)
     * 1. UI 개선
     * 2. 추출할 모드를 우클릭 후, 관련된 모든 모드를 참조모드로 선택 가능
     */

3빼고 다 구현 완료했습니다~

메인 창에서 참조 모드 표시가 하나만 표시될텐데 더 많이 표시할 수 있도록 작업 중입니다.

한가지 궁금한게있는데요, 프로그램을 실행할때 추가적으로 설치해야 하는 것이 있었나요?
.net runtime 7.0 <- 이거 따로 설치해야 프로그램이 실행되었나요?

**Reply by csh1668** [reply]
이쯤이면 배포해도 되겠다 싶어서 repo 하나 팠어요
https://github.com/csh1668/RimworldExtractor
readme나 경고/에러에 대한 설명을 더 추가하고 배포하려고 합니다

**Reply by RimWorldKorea** [reply]
마참내!

제가 당장 테스트할 여건이 안돼서 다음 주말동안 천천히 써봐야겠어요

저는 옛날에 닷넷 런타임을 따로 깔긴 했는데
닷넷도 같이 포함시켜서 컴파일한거 아니면 없을땐 따로 깔아줘야 하지 않아요?

**Reply by csh1668** [reply]
같이 포함시킬 수가 있었군요..!

https://github.com/csh1668/RimworldExtractor/releases
이걸 사용해주세요

피드백은 당분간은 RMK쪽 채널에서 부탁드리겠습니다 감사합니다

**Reply by zilrty** [reply]
선생님 이제 알파 버리고 갈아타도 되는 부분입니까

**Comment by RimWorldKorea** [comment]
-> #496
